localstack/aws/protocol/parser.py

```diff
@@ -936,6 +1009,111 @@ def _get_serialized_name(self, shape: Shape, default_name: str) -> str:
             return default_name
 
 
+class S3RequestParser(RestXMLRequestParser):
+    def parse(self, request: HttpRequest) -> Tuple[OperationModel, Any]:
+        """Handle virtual-host-addressing for S3."""
+        if (
+            # TODO implement a more sophisticated determination if the host contains S3 virtual host addressing
+            not request.host.startswith("s3.")
+            and not request.host.startswith("localhost.")
+            and not request.host.startswith("127.0.0.1")
+        ):
+            self._revert_virtual_host_style(request)
+        return super().parse(request)
+
+    def _revert_virtual_host_style(self, request: HttpRequest):
+        # extract the bucket name from the host part of the request
+        bucket_name = request.host.split(".")[0]
+        # split the url and put the bucket name at the front
+        parts = urlsplit(request.url)
+        path_parts = parts.path.split("/")
+        path_parts = [bucket_name] + path_parts
+        path_parts = [part for part in path_parts if part]
+        path = "/" + "/".join(path_parts) or "/"
+        # set the path with the bucket name in the front at the request
+        request.path = path
```

Comment:
we'll have to see whether this causes issues with the new handler chain framework, where the request object is passed through multiple handlers. modifying fundamental aspects of the request, such as the path, may cause unexpected behavior down the chain. cloning the request at line 1021 and handing it down for parsing would probably be better. but there's no good clone method yet, so let's just leave it for now.

Reply:
I thought of that as well. I'll add a TODO and we can tackle this later on. This code is only used for S3 (so it isn't used yet at all).